fastwarc/fastwarc/stream_io.pyi

```diff
@@ -36,28 +53,28 @@ class CompressingStream(IOStream):
 
 class BrotliStream(CompressingStream):
     def __init__(
-        self, raw_stream: IOStream, quality: int = 11, lgwin: int = 22, lgblock: int = 0
+        self, raw_stream: Union[IOStream, BinaryIO, _ReadableStream, _WritableStream], quality: int = 11, lgwin: int = 22, lgblock: int = 0
     ) -> None: ...
 
 
 class GZipStream(CompressingStream):
     def __init__(
-        self, raw_stream: IOStream, compression_level: int = 9, zlib: bool = False
+        self, raw_stream: Union[IOStream, BinaryIO, _ReadableStream, _WritableStream], compression_level: int = 9, zlib: bool = False
```

Comment:
One more question: Do we need `_ReadableStream` and `_WritableStream`? Isn't `IOStream` enough?

Reply:
>We need at least close() and tell()

Thanks, I missed that!

I'm very happy to include `close` and `tell` as one of the required methods, I'll push a commit shortly.

In the case of `CompressingStream`s, `close` seems indeed to be called at `__dealloc__` while `tell` is called during `__cinit__`. There is one explicit call to `tell` in the `fastwarc.warc.ArchiveIterator` and `close` will probably called implicitly also through this `__dealloc__` method.

Possibly one should include a test for that, i.e. providing a minimal class that implements these, there are already tests that test with generic `io.BytesIO` and `open` objects.
 

>and at that point you are already at IOStream.

Do you mean `fastwarc.stream_io.IOStream` or a class that satisfies `typing.BinaryIO`? If `IOStream`: I would disagree with that, since the point of the `Protocol`s is to allow any arbitrary python class (as the Cython code does), i.e. *not* having to *inherit* from `IOStream`.

As a datapoint that possibly helps you understanding my standpoint: We're using `fastwarc` to a broad degree, but more for interacting with WARC files on already decompressed, custom streams. We are not using any of the stream classes of this package.

Reply:
I was thinking of something like `Union[IOStream, BinaryIO]`. The former is satisfied by FastWARC's internal `stream_io` classes, the other is any other Python file-like object that would be wrapped in `PythonIOStreamAdapter`.

Reply:
Think more about it, I would prefer `Union[IOStream, BinaryIO]`. It's stricter and therefore more future-safe.

Reply:
One more thing: All members of `WarcRecordType` should be available as namespace-less flags, not juts `no_type` and `any_type`.

Reply:
>If you want laxer duck typing, you could make IOStream a Protocol class via multi-inheritance.

This feels messy to me, IMHO.

>Thinking more about it, I would prefer Union[IOStream, BinaryIO]. It's stricter and therefore more future-safe.

Oh no, then maybe my point did not come across correctly. Doing it this way would break everything around our approach and not helping. 

Just take a look at `gzip.GzipFile` for example, just picking *something* from a battle-tested, well maintainted upstream library: https://github.com/python/cpython/blob/253c6a0b2f88b3327b7113860b99d665346fe43c/Lib/gzip.py#L142
This class *does* very much implement everything that's required for everything in `fastwarc`, but still *fails* a check against *typing.BinaryIO*. This is because it's missing `truncate()`, although it's nowhere used in this code.

Moreover, the entire point of `PythonIOStreamAdapter` is that it can wrap arbitrary classes, as long as they implement the minimal set of required methods. Forcing a `typing.BinaryIO` here would somewhat nullify that.

I'd *strongly* recommend leaving in the proper `Protocol` classes here, as they perfectly map on to the logic of this code and also would nullify the advantage that this PR has.

Reply:
>One more thing: All members of WarcRecordType should be available as namespace-less flags, not juts no_type and any_type.

Sure, can do, but our of curiosity: What is the value of exposing every possible value of this enum as a separate variable at top-level? Is this something that has been done for backwards compatibility?

Reply:
> This class does very much implement everything that's required for everything in fastwarc, but still fails a check against typing.BinaryIO. This is because it's missing truncate(), although it's nowhere used in this code.
> 
> Moreover, the entire point of PythonIOStreamAdapter is that it can wrap arbitrary classes, as long as they implement the minimal set of required methods. Forcing a typing.BinaryIO here would somewhat nullify that

Technically, the API documentation allows only "file-like" objects, so they should implement everything that's needed. But fair point. I'm fine with having a `Protocol` class then, but it shouldn't be split into readable and writable, since the interface doesn't guarantee that a read-only or write-only stream would be safe. You could technically pass a read-only stream to `ArchiveIterator`, but everything else must be read-write and should require all methods that `IOStream` has.

> Sure, can do, but our of curiosity: What is the value of exposing every possible value of this enum as a separate variable at top-level? Is this something that has been done for backwards compatibility?

Not really for compatibility, but for ease of use. `record.record_type == WarcRecordType.continuation` is a rather unwieldy expression and checking for record types is a pretty common action.

Reply:
Went for a general `_GenericIOStream` protocol. 

Exposed all enum values at top level.